### PR TITLE
Updated Exceptions for SIM step

### PR DESCRIPTION
**Pull request by civanch**
#### PR description:
In this PR unification of Exception scheme for SIM step is provided. During initialization cms::Exception is used. In run time G4Exception instead. G4Exception is customized by ExceptionHandler provided in CMSSW, which is using cms::Exception. For debug purpose a user may enable Geant4 trace if g4SimHits variable Trace=True.

This PR should not change results successful WFs but may be useful for debugging of SIM problems.

#### PR validation:
private

### CODE REVIEW

**Comment by civanch** [comment]
please test

**Comment by civanch** [comment]
+1

**Comment by perrotta** [comment]
+1

File: SimG4Core/Notification/src/GenParticleInfoExtractor.cc [hunk 1]
```diff
@@ -1,14 +1,18 @@
 #include "SimG4Core/Notification/interface/GenParticleInfoExtractor.h"
-#include "SimG4Core/Notification/interface/SimG4Exception.h"
-
 #include "G4PrimaryParticle.hh"
 
 const GenParticleInfo &GenParticleInfoExtractor::operator()(const G4PrimaryParticle *p) const {
   G4VUserPrimaryParticleInformation *up = p->GetUserInformation();
   if (up == nullptr)
-    throw SimG4Exception("GenParticleInfoExtractor: G4PrimaryParticle has no user information");
+    G4Exception("SimG4Core/Notification",
```
Comment (perrotta) [inline SimG4Core/Notification/src/GenParticleInfoExtractor.cc:7]: Uhm, I was too fast in signing this before inspecting the static analyzer report.

In the previous code if `up == nullptr` we were throwing here. Now, there is this G4Exception (is it also throwing?), and therefore a null reference can possibly be returned by this method. Is it this the case? Is it the expected/wanted behaviour?

File: SimG4Core/Notification/src/TrackInformationExtractor.cc [hunk 2]
```diff
@@ -21,3 +21,13 @@ TrackInformation &TrackInformationExtractor::operator()(G4Track &gtk) const {
     wrongType();
   return *tkInfo;
 }
+
+void TrackInformationExtractor::missing(const G4Track &) const {
+  G4Exception(
```
Comment (perrotta) [inline SimG4Core/Notification/src/TrackInformationExtractor.cc:26]: Same here: not throwing here would make `operator()` return a null pointer

File: SimG4Core/Notification/src/TrackInformationExtractor.cc [hunk 3]
```diff
@@ -21,3 +21,13 @@ TrackInformation &TrackInformationExtractor::operator()(G4Track &gtk) const {
...
+      "SimG4Core/Notification", "mc001", FatalException, "TrackInformationExtractor: G4Track has no TrackInformation");
+}
+
+void TrackInformationExtractor::wrongType() const {
+  G4Exception(
```
Comment (perrotta) [inline SimG4Core/Notification/src/TrackInformationExtractor.cc:31]: Same here: not throwing here would make `operator()` return a null pointer

**Review by perrotta — COMMENTED** [review]
@civanch sorry for the late reaction of mine, but could you please comment about what below (and also provide a fix, if you deem it needed)?

**Comment by civanch** [comment]
@perrotta , I will do the next PR. It is a typical situation with static analyzer and G4Exception - G4 throwing an exception but the static analyzer do not know about. With simple tricks this may be avoided.

**Pull request merged**